File: LeetCode/56. Merge Intervals.py
# 思路比较简单，按起始点排序，然后从前往后遍历，终止点能连上就合并，不能就储存起来，用下一个开始继续查询
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if (len(intervals) < 2):
            return intervals
        intervals.sort(key=lambda x:x[0])
        current = None
        result = []
        for pair in intervals:
            if (not current):
                current = pair
                continue
            if (current[1] >= pair[0]):
                current[1] = max(current[1], pair[1])
            else:
                result.append(current)
                current = pair
        result.append(current)
        return result

File: LeetCode/56. Merge Intervals.cpp
// 这题的解法很简单，先按开始区间排序，然后把前后的合并放进新的vector，遍历处理完所有区间即可，注意边界条件处理就行
bool comp(vector<int>& x, vector<int>& y) {
    return x[0] < y[0];
}
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.size() == 0) {
            return intervals;
        }
        sort(intervals.begin(), intervals.end(), comp);
        vector<vector<int>> res = vector<vector<int>>();
        vector<int> cur = intervals[0];
        for (int i = 0; i < intervals.size(); ++i) {
            if (intervals[i][0] > cur[1]) {
                res.push_back(cur);
                cur = intervals[i];
            }
            else {
                cur[0] = min(cur[0], intervals[i][0]);
                cur[1] = max(cur[1], intervals[i][1]);
            }
        }
        res.push_back(cur);
        return res;
    }
    
};